### PR TITLE
[WIP] Add debiased barycenter (Sinkhorn + convolutional sinkhorn)

### PR DESCRIPTION
**Pull request by hichamjanati**
## Types of changes

- [x] Add debiased barycenter 
- [x] Add debiased convolutional barycenter
- [x] Add convolutional barycenter wrapper
- [x] Test "debiased" arg
- [x] Remove while loops -> for loops + break
- [x] Example for debiased barycenters

### CODE REVIEW

**Comment by hichamjanati** [comment]
Jax doesn't support item assignment in arrays which is needed in sinkhorn_log functions  (for e.g here https://github.com/hichamjanati/POT/blob/5297495058d662a5f4eee779c5b33dd20343cb30/ot/bregman.py#L1499) to avoid using lists. 

any obvious workaround for this ? I'm thinking adding a method `nx.assign` that does a normal subscription with np and torch and the specific .at().set() for jax. WDYT ?

**Comment by rflamary** [comment]
barycenters are done with a relatively small number of distributions in practice so using lists is not much of a problem here (you can stack after the loop). I agree this is a pain so either you use list (maybe test to see the overhead in practice but it should be small) or you raise a not implemented for jax.

File: examples/barycenters/plot_barycenter_1D.py [hunk 1]
```diff
@@ -56,11 +56,11 @@
 
 #%% plot the distributions
 
-pl.figure(1, figsize=(6.4, 3))
-for i in range(n_distributions):
-    pl.plot(x, A[:, i])
-pl.title('Distributions')
-pl.tight_layout()
+# plt.figure(1, figsize=(6.4, 3))
```
[the remark this answers is not in the document]
Comment (rflamary) [reply examples/barycenters/plot_barycenter_1D.py:59]: # sphinx_gallery_thumbnail_number = 4

here choose the number of the last figure s that one can see it in the gallery

File: examples/barycenters/plot_debiased_barycenter.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+.. [28] Janati, H., Cuturi, M., Gramfort, A. Proceedings of the 37th
+        International Conference on Machine Learning, PMLR 119:4692-4701, 2020
+"""
+
+# Author: Hicham Janati <hicham.janati100@gmail.com>
+#
+# License: MIT License
+
+# sphinx_gallery_thumbnail_number = 4
```
Comment (rflamary) [inline examples/barycenters/plot_debiased_barycenter.py:19]: pick a number that exists for the gallery

File: ot/bregman.py [hunk 3]
```diff
@@ -116,11 +119,14 @@ def sinkhorn(a, b, M, reg, method='sinkhorn', numItermax=1000,
     References
     ----------
 
-    .. [2] M. Cuturi, Sinkhorn Distances : Lightspeed Computation of Optimal Transport, Advances in Neural Information Processing Systems (NIPS) 26, 2013
+    .. [2] M. Cuturi, Sinkhorn Distances : Lightspeed Computation of Optimal Transport,
```
Comment (rflamary) [inline ot/bregman.py:122]: this breaks the referenvce in the doc. either keep the long lie or fid a way to have references that look OK

File: ot/bregman.py [hunk 4]
```diff
@@ -250,27 +262,31 @@ def sinkhorn2(a, b, M, reg, method='sinkhorn', numItermax=1000,
     References
     ----------
 
-    .. [2] M. Cuturi, Sinkhorn Distances : Lightspeed Computation of Optimal Transport, Advances in Neural Information Processing Systems (NIPS) 26, 2013
+    .. [2] M. Cuturi, Sinkhorn Distances : Lightspeed Computation of
```
Comment (rflamary) [inline ot/bregman.py:265]: here again

File: ot/bregman.py [hunk 5]
```diff
@@ -382,7 +402,8 @@ def sinkhorn_knopp(a, b, M, reg, numItermax=1000,
     References
     ----------
 
-    .. [2] M. Cuturi, Sinkhorn Distances : Lightspeed Computation of Optimal Transport, Advances in Neural Information Processing Systems (NIPS) 26, 2013
+    .. [2] M. Cuturi, Sinkhorn Distances : Lightspeed Computation
```
Comment (rflamary) [inline ot/bregman.py:405]: again :
image

File: ot/bregman.py [hunk 6]
```diff
@@ -455,13 +476,19 @@ def sinkhorn_knopp(a, b, M, reg, numItermax=1000,
...
             if verbose:
-                if cpt % 200 == 0:
+                if ii % 200 == 0:
                     print(
                         '{:5s}|{:12s}'.format('It.', 'Err') + '\n' + '-' * 19)
-                print('{:5d}|{:8e}|'.format(cpt, err))
-        cpt = cpt + 1
+                print('{:5d}|{:8e}|'.format(ii, err))
+    else:
+        warnings.warn("Sinkhorn did not converge. You might want to "
```
Comment (rflamary) [inline ot/bregman.py:487]: print warning only if verbose ?

Comment (rflamary) [reply ot/bregman.py:487]: i mean it never conveges for small egularizaytion and you don't want warning at each iteration in your gradient descent ;)

File: ot/bregman.py [hunk 7]
```diff
@@ -612,7 +639,7 @@ def sinkhorn_log(a, b, M, reg, numItermax=1000,
         if log:
             log = {'err': []}
 
-        Mr = M / (-reg)
+        Mr = - M / reg
```
Comment (rflamary) [inline ot/bregman.py:642]: Keep the original, it is more efficient
image

File: ot/bregman.py [hunk 8]
```diff
@@ -647,13 +673,19 @@ def get_logT(u, v):
                     log['err'].append(err)
 
                 if verbose:
-                    if cpt % 200 == 0:
+                    if ii % 200 == 0:
                         print(
                             '{:5s}|{:12s}'.format('It.', 'Err') + '\n' + '-' * 19)
-                    print('{:5d}|{:8e}|'.format(cpt, err))
-            cpt = cpt + 1
+                    print('{:5d}|{:8e}|'.format(ii, err))
```
Comment (rflamary) [inline ot/bregman.py:679]: add verbose in est to check that nothing breaks please

File: ot/bregman.py [hunk 9]
```diff
@@ -647,13 +673,19 @@ def get_logT(u, v):
...
+                if err < stopThr:
+                    break
+        else:
+            warnings.warn("Sinkhorn did not converge. You might want to "
+                          "increase the number of iterations `numItermax` "
```
Comment (rflamary) [inline ot/bregman.py:684]: warning only if verbose + test wih call to functio with numItermax=1

File: ot/bregman.py [hunk 10]
```diff
@@ -730,9 +767,12 @@ def greenkhorn(a, b, M, reg, numItermax=10000, stopThr=1e-9, verbose=False,
     References
     ----------
 
-    .. [2] M. Cuturi, Sinkhorn Distances : Lightspeed Computation of Optimal Transport, Advances in Neural Information Processing Systems (NIPS) 26, 2013
+    .. [2] M. Cuturi, Sinkhorn Distances : Lightspeed Computation of Optimal Transport,
```
Comment (rflamary) [inline ot/bregman.py:770]: broken reference

File: ot/bregman.py [hunk 11]
```diff
@@ -794,14 +835,16 @@ def greenkhorn(a, b, M, reg, numItermax=10000, stopThr=1e-9, verbose=False,
             viol += (-old_v + new_v) * K[:, i_2] * u
             viol_2[i_2] = new_v * K[:, i_2].dot(u) - b[i_2]
             v[i_2] = new_v
-            # print('b',np.max(abs(aviol -viol)),np.max(abs(aviol_2 - viol_2)))
 
         if stopThr_val <= stopThr:
             break
     else:
-        print('Warning: Algorithm did not converge')
+        warnings.warn("Sinkhorn did not converge. You might want to "
```
Comment (rflamary) [inline ot/bregman.py:842]: verbose only

File: ot/bregman.py [hunk 12]
```diff
@@ -885,11 +933,15 @@ def sinkhorn_stabilized(a, b, M, reg, numItermax=1000, tau=1e3, stopThr=1e-9,
     References
     ----------
 
-    .. [2] M. Cuturi, Sinkhorn Distances : Lightspeed Computation of Optimal Transport, Advances in Neural Information Processing Systems (NIPS) 26, 2013
+    .. [2] M. Cuturi, Sinkhorn Distances : Lightspeed Computation of
```
Comment (rflamary) [inline ot/bregman.py:936]: broken references

File: ot/bregman.py [hunk 13]
```diff
@@ -991,34 +1040,35 @@ def get_Gamma(alpha, beta, u, v):
...
-            print('Warning: numerical errors at iteration', cpt)
+            warnings.warn('Numerical errors at iteration', ii)
             u = uprev
             v = vprev
             break
-
-        cpt = cpt + 1
-
+    else:
+        warnings.warn("Sinkhorn did not converge. You might want to "
```
Comment (rflamary) [inline ot/bregman.py:1062]: verbose

**Review by rflamary — COMMENTED** [review]
This is great but we have a bade coverage due to all the parameters. could you put verbose=True on some test and also test for the warning when numItermax is very small?
